### PR TITLE
Add OpenSearchCatalog to enable direct access OpenSearch index in Spark SQL

### PR DESCRIPTION
**Pull request by penghuo**
### Description
This PR introduces the `OpenSearchCatalog` class, enabling Apache Spark to integrate with OpenSearch as a data catalog and direct access OpenSearch Index in Spark SQL. Key features and changes include:

### Configuration Parameters
- **`opensearch.port`**: Default is `9200`.
- **`opensearch.scheme`**: Default is `http`. Valid values are `[http, https]`.
- **`opensearch.auth`**: Default is `noauth`. Valid values are `[noauth, sigv4, basic]`.
- **`opensearch.auth.username`**: Basic auth username.
- **`opensearch.auth.password`**: Basic auth password.
- **`opensearch.region`**: Default is `us-west-2`. Used only when `auth` is `sigv4`.

### Implemented Methods
- **`loadTable`**: Loads a read only table from the OpenSearch index.
- **Unsupported Operations**: The following methods throw `UnsupportedOperationException`:
  - `listTables`
  - `createTable`
  - `alterTable`
  - `dropTable`
  - `renameTable`

### Follow up
* Support comma separated index name as table identifier.

### Doc
https://github.com/penghuo/penghuo-opensearch-spark/blob/issue395/docs/opensearch-table.md

### Issues Resolved
https://github.com/opensearch-project/opensearch-spark/issues/395

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: integ-test/src/test/scala/org/apache/spark/opensearch/catalog/OpenSearchCatalogITSuite.scala [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+import org.apache.spark.FlintSuite
+import org.apache.spark.sql.{QueryTest, Row}
+import org.apache.spark.sql.streaming.StreamTest
+
+class OpenSearchCatalogITSuite
+    extends QueryTest
+    with StreamTest
+    with FlintSuite
+    with OpenSearchSuite {
```
Comment (dai-chen) [inline integ-test/src/test/scala/org/apache/spark/opensearch/catalog/OpenSearchCatalogITSuite.scala:18]: np: use `FlintSparkSuite`?

File: docs/opensearch-table.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+```
+
+Using a wildcard index name:
+```scala
+val df = spark.sql("SELECT * FROM dev.default.`my_index*`")
+df.show()
+```
+
+## Limitation
+### Unsupported catalog operation
```
Comment (dai-chen) [inline docs/opensearch-table.md:52]: Does `DESC` table work?
With current implementation, is `INSERT`, `UPDATE`, `DELETE` supported or not?

**Review by dai-chen — APPROVED** [review]

File: docs/opensearch-table.md (hunk 2 above)
Comment (seankao-az) [reply docs/opensearch-table.md:52]: +1 on this question.
FlintTable supports write, so INSERT might work. Not really sure if our implementation supports UPDATE and DELETE

Comment (penghuo) [reply docs/opensearch-table.md]: Yes, INSERT is support. Change the Table as read only table, the reasons are (1) read is major use case. (2) OpenSearch does not support wildcard index pattern / comma seperated index write operation. 

Add IT to cover INSERT, UPDATE, DELETE.

in summary
* OpenSearch index direct access is read only.
* OpenSearch table will support write operation.

File: integ-test/src/test/scala/org/apache/spark/opensearch/catalog/OpenSearchCatalogITSuite.scala (hunk 1 above)
Comment (penghuo) [reply integ-test/src/test/scala/org/apache/spark/opensearch/catalog/OpenSearchCatalogITSuite.scala]: done.

File: flint-spark-integration/src/main/scala/org/apache/spark/sql/flint/FlintReadOnlyTable.scala [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        .values()
+        .asScala
+        .headOption
+        .map(m => FlintDataType.deserialize(m.getContent))
+        .getOrElse(StructType(Nil))
+    }
+  }
+
+  override def capabilities(): util.Set[TableCapability] =
+    util.EnumSet.of(BATCH_READ, BATCH_WRITE, TRUNCATE, STREAMING_WRITE)
```
Comment (seankao-az) [inline flint-spark-integration/src/main/scala/org/apache/spark/sql/flint/FlintReadOnlyTable.scala:58]: `BATCH_WRITE` and `STREAMING_WRITE` should be moved to `FlintTable`

File: flint-spark-integration/src/main/scala/org/apache/spark/opensearch/catalog/OpenSearchCatalog.scala [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    this.catalogName = name
+    this.options = options
+  }
+
+  override def name(): String = catalogName
+
+  @throws[NoSuchNamespaceException]
+  override def listTables(namespace: Array[String]): Array[Identifier] = {
+    throw new UnsupportedOperationException("OpenSearchCatalog does not support listTables")
+  }
```
Comment (seankao-az) [inline flint-spark-integration/src/main/scala/org/apache/spark/opensearch/catalog/OpenSearchCatalog.scala:45]: what does `@throws` annotation do here? Is it required
It's throwing a different exception

File: flint-spark-integration/src/main/scala/org/apache/spark/sql/flint/FlintReadOnlyTable.scala (hunk 3 above)
Comment (penghuo) [reply flint-spark-integration/src/main/scala/org/apache/spark/sql/flint/FlintReadOnlyTable.scala:60]: Thx

**Pull request merged**